packages/dev/core/src/Collisions/gpuPicker.ts

```diff
@@ -0,0 +1,314 @@
+                    const mesh = this._pickableMeshes[index];
+                    if (mesh.hasInstances) {
+                        (mesh as Mesh).removeVerticesData(VertexBuffer.ColorKind);
+                    }
+                }
+            }
+            this._userDefinedList = false;
+            this._pickableMeshes = [];
+            this._idMap = {};
+            this._idColors = [];
+            return;
+        }
+        this._userDefinedList = true;
+        this._pickableMeshes = list;
+
+        // We will affect colors and create vertex color buffers
+        let id = 1;
+        for (let index = 0; index < this._pickableMeshes.length; index++) {
+            const mesh = this._pickableMeshes[index];
+            mesh.useVertexColors = false;
+
+            if (mesh.isAnInstance) {
+                continue; // This will be handled by the source mesh
+            }
+
+            const r = (id & 0xff0000) >> 16;
+            const g = (id & 0x00ff00) >> 8;
+            const b = (id & 0x0000ff) >> 0;
+            this._idMap[`${r}_${g}_${b}`] = index;
+            id++;
+
+            if (mesh.hasInstances) {
+                const instances = (mesh as Mesh).instances;
+                const colorData = new Float32Array(4 * (instances.length + 1));
+                const engine = mesh.getEngine();
+
+                colorData[0] = r / 255.0;
+                colorData[1] = g / 255.0;
+                colorData[2] = b / 255.0;
+                colorData[3] = 1.0;
+                for (let i = 0; i < instances.length; i++) {
+                    const instance = instances[i];
+                    const r = (id & 0xff0000) >> 16;
+                    const g = (id & 0x00ff00) >> 8;
+                    const b = (id & 0x0000ff) >> 0;
+                    this._idMap[`${r}_${g}_${b}`] = this._pickableMeshes.indexOf(instance);
+
+                    colorData[(i + 1) * 4] = r / 255.0;
+                    colorData[(i + 1) * 4 + 1] = g / 255.0;
+                    colorData[(i + 1) * 4 + 2] = b / 255.0;
+                    colorData[(i + 1) * 4 + 3] = 1.0;
+                    id++;
+                }
+
+                const buffer = new VertexBuffer(engine, colorData, VertexBuffer.ColorKind, false, false, 4, true);
+                (mesh as Mesh).setVerticesBuffer(buffer, true);
+            } else {
+                this._idColors[mesh.uniqueId] = Color3.FromInts(r, g, b);
+            }
+        }
+    }
+
+    /**
+     * Execute a picking operation
+     * @param x defines the X coordinates where to run the pick
+     * @param y defines the Y coordinates where to run the pick
+     * @param scene defines the scene to pick from
+     * @param disposeWhenDone defines a boolean indicating we do not want to keep resources alive
+     * @returns A promise with the picking results
+     */
+    public pickAsync(x: number, y: number, scene: Scene, disposeWhenDone = true): Promise<Nullable<AbstractMesh>> {

```

Comment:
I'm not sure disposing after a single pick has been done is the most common use case? If not, I would set `disposeWhenDone` to `false` by default.